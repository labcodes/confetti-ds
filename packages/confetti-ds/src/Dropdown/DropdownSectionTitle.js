import React, { Component } from "react";
import { PropTypes } from "prop-types";

export default class DropdownSectionTitle extends Component {
  static propTypes = {
    /** This is DropdownSectionTitle's title */
    text: PropTypes.node.isRequired,
    /** This is DropdownSectionTitle's color. The color is inherit from dropdown color  */
    color: PropTypes.string,
  };

  static defaultProps = {
    color: "teal",
  };

  render() {
    const { text, color } = this.props;
    return (
      <span
        className={`lab-dropdown__section-title lab-dropdown__section-title--color-${color}`}
      >
        {text}
      </span>
    );
  }
}
