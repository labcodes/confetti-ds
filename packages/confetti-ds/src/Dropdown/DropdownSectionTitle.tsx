import React from "react";

export interface DropdownSectionTitleProps {
  /** This is DropdownSectionTitle's title */
  text: React.ReactNode;
  /** This is DropdownSectionTitle's color. The color is inherit from dropdown color  */
  color?: "white" | "mineral" | "teal" | "purple";
}
export default function DropdownSectionTitle({
  text,
  color = "teal",
}: DropdownSectionTitleProps) {
  return (
    <span
      className={`lab-dropdown__section-title lab-dropdown__section-title--color-${color}`}
    >
      {text}
    </span>
  );
}
