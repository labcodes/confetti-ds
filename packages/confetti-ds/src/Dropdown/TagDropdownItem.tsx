import React, {
  SyntheticEvent,
  useEffect,
  useCallback,
  MutableRefObject,
} from "react";
import { isEmpty } from "lodash";

import Icon from "../Icon";
import { IconTypes, TagTypes } from "../constants";
import AbstractTag from "../Tags/AbstractTag";

export interface BaseTagDropdownItemProps {
  /** This is the Tag's text. */
  text: string;
  /** Disables the Tag. Won't be read by screen readers. */
  disabled?: boolean;
  /** Action to be executed when the Tag is clicked. */
  onClick?: (event: SyntheticEvent) => any;
  /** Type of the icon to be rendered. Won't render an icon if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
  icon?: IconTypes;
  /** Sets Tag's color. */
  color?: TagTypes;
  /** Skin of the rendered Tag. */
  skin?: "pale" | "vivid";
  /** Sets an outline style. */
  isOutline?: boolean;
  /** Option tab index */
  tabIndex?: number;
  /** This function is used on AbstractTag's componenentDidMount to set the current Ref */
  setRef?: (ref: MutableRefObject<HTMLButtonElement>) => any;
  /** This function is used to handle click or keydown interactions */
  onInteraction?: ({
    event,
    ref,
  }: {
    event: any;
    ref: MutableRefObject<boolean>;
  }) => any;
  /** This is the option's value */
  value?: string;
  /** This prop is a boolean to verify if the option is current selected  */
  isSelected?: boolean;
}

export type TagDropdownItemProps =
  | (BaseTagDropdownItemProps & {
      /** Source of the thumb to be rendered. Won't render a thumb if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
      thumbSrc?: undefined;
      /** Alt text of the thumb to be rendered. */
      thumbAlt?: undefined;
    })
  | (BaseTagDropdownItemProps & {
      /** Source of the thumb to be rendered. Won't render a thumb if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
      thumbSrc: string;
      /** Alt text of the thumb to be rendered. */
      thumbAlt: string;
    });

export default function TagDropdownItem({
  text,
  disabled = false,
  onClick = () => {},
  icon,
  color,
  skin = "pale",
  isOutline = false,
  tabIndex = 0,
  setRef = () => {},
  onInteraction = () => {},
  value,
  isSelected = false,
  thumbSrc,
  thumbAlt,
}: TagDropdownItemProps) {
  const renderThumb = () =>
    thumbSrc ? (
      <img className="lab-tag__thumb" src={thumbSrc} alt={thumbAlt} />
    ) : undefined;

  const renderIcon = () =>
    icon ? (
      <Icon type={icon} color="black-75" className="lab-tag--left-icon" />
    ) : undefined;

  // super constructor
  const checkThumbAndIcon = useCallback(() => {
    const errorMessage =
      "`TagDropdownItem` can't be initialized with both `thumb` and `icon` props.";
    if (!isEmpty(thumbSrc) && !isEmpty(icon)) {
      throw new Error(errorMessage);
    }
  }, [thumbSrc, icon]);

  // componentDidUpdate
  useEffect(() => {
    checkThumbAndIcon();
  }, [thumbSrc, icon]);

  return (
    <AbstractTag
      text={text}
      className={`lab-tag--togglable${icon ? ` lab-tag--has-left-icon` : ""}${
        thumbSrc ? ` lab-tag--has-thumb` : ""
      }`}
      isOutline={isOutline}
      skin={skin}
      color={color}
      renderPrefix={renderThumb() || renderIcon()}
      onClick={onClick}
      disabled={disabled}
      tabIndex={tabIndex}
      onInteraction={onInteraction}
      isDropdown
      setRef={setRef}
      value={value}
      isSelected={isSelected}
      role="option"
    />
  );
}
