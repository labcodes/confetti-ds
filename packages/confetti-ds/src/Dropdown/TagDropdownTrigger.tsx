import React, {
  Dispatch,
  MutableRefObject,
  SetStateAction,
  SyntheticEvent,
} from "react";
import AbstractTag from "../Tags/AbstractTag";
import { IconTypes, ICON_TYPES } from "../constants";
import Icon from "../Icon";

interface TagDropdownTriggerProps {
  /** This is the Tag's text. */
  text: string;
  /** Type of the icon to be rendered. Won't render an icon if not passed to the component. */
  icon?: IconTypes;
  /** Sets Tag's color. */
  color?: string;
  /** Skin of the rendered Tag. */
  skin?: string;
  /** Sets an outline style. */
  isOutline?: boolean;
  /** Disables the Tag. Will be read by screen readers. When true, will override `disabled`. */
  ariaDisabled?: boolean;
  /** Disables the Tag. Won't be read by screen readers. */
  disabled?: boolean;
  /** tabIndex is used to define the navigation order for focusable elements. If not undefined, it is passed to the tag component. */
  tabIndex?: number;
  /** This function is used to handle click or keydown interactions */
  onInteraction?: ({ event }: { event: SyntheticEvent }) => any;
  /** This function is used on AbstractTag to set the current Ref */
  setRef?: Dispatch<SetStateAction<MutableRefObject<HTMLButtonElement>>>;
}

export default function TagDropdownTrigger({
  text,
  icon,
  color,
  skin = "pale",
  isOutline = false,
  ariaDisabled = false,
  disabled = false,
  tabIndex,
  onInteraction = () => {},
  setRef = () => {},
}: TagDropdownTriggerProps) {
  // eslint-disable-next-line react/display-name
  const handleIcon = () =>
    icon ? (
      <Icon
        type={icon}
        color="black-75"
        size="petit"
        className="lab-tag--left-icon"
      />
    ) : undefined;

  const dropdownIcon = () => (
    <span className="lab-tag__dropdown-icon-wrapper">
      <Icon
        type="dropdown-closed"
        color="black-75"
        size="petit"
        className="lab-tag--dropdown-icon"
      />
    </span>
  );

  return (
    <AbstractTag
      className={`lab-tag--dropdown${icon ? ` lab-tag--has-left-icon` : ""}`}
      text={text}
      color={color}
      skin={skin}
      isOutline={isOutline}
      disabled={!ariaDisabled && disabled}
      ariaDisabled={ariaDisabled}
      renderPrefix={handleIcon()}
      renderSuffix={dropdownIcon()}
      tabIndex={tabIndex}
      onInteraction={onInteraction}
      isDropdown
      setRef={setRef}
      role="listbox"
    />
  );
}
