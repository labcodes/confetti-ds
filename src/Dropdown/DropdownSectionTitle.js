import React, { Component } from "react";
import { PropTypes } from "prop-types";

export default class DropdownSectionTitle extends Component {
  static propTypes = {
    text: PropTypes.node.isRequired,
    color: PropTypes.string,
  };

  static defaultProps = {
    color: "teal",
  };

  render() {
    const { text, color } = this.props;
    return (
      <span
        className={`lab-dropdown__section-title lab-dropdown__section-title--color-${color}`}
      >
        {text}
      </span>
    );
  }
}
