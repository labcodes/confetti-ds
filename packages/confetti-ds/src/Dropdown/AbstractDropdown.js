import React, { Component } from "react";
import PropTypes from "prop-types";

import { DropdownTag, TagItem } from "../Tags";
import { Button } from "../Button";
import { dropdownOptions, expectedKeys } from "./propTypes";
import DropdownSectionTitle from "./DropdownSectionTitle";
import DropdownTrigger from "./DropdownTrigger";
import DropdownOption from "./DropdownOption";

export default class AbstractDropdown extends Component {
  static propTypes = {
    // functions
    onOpen: PropTypes.func,
    onClose: PropTypes.func,
    onSelect: PropTypes.func.isRequired,
    // props
    dropdownType: PropTypes.oneOf(dropdownOptions.types).isRequired,
    color: PropTypes.string,
    defaultText: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired,
    id: PropTypes.string.isRequired,
  };

  static defaultProps = {
    // props
    color: "teal",
    // functions
    onOpen: () => {},
    onClose: () => {},
  };

  constructor(props) {
    super(props);
    this.state = {
      isOpen: false,
      refList: [],
      selected: { ref: null, index: null },
      focused: { index: null },
      displayText: props.defaultText,
    };
  }

  componentDidMount() {
    const { checkIfHasChildren } = this;

    checkIfHasChildren();
  }

  componentDidUpdate(prevProps, prevState) {
    const { triggerRef, checkIfHasChildren } = this;

    const { isOpen, selected, refList } = this.state;

    const { onClose, onOpen } = this.props;

    if (isOpen !== prevState.isOpen) {
      if (isOpen === false) {
        triggerRef.current.focus();

        onClose();
        return;
      }

      const hasSelectedOption = selected.ref;
      if (hasSelectedOption) selected.ref.current.focus();
      else {
        const firstOption = refList[0];
        if (firstOption) firstOption.current.focus();
      }

      onOpen();
    }

    checkIfHasChildren();
  }

  checkIfHasChildren = () => {
    const { children } = this.props;

    const hasNoChildren = !React.Children.count(children);
    if (hasNoChildren)
      throw Error("You need to pass children to the dropdown component.");
  };

  onClickEvent = () => {
    const { isOpen, selected } = this.state;
    const focusedIndex = this.getSelectedOptionIndex(selected);
    this.setState({ isOpen: !isOpen, focused: { index: focusedIndex } });
  };

  getSelectedOptionIndex = (selected) => {
    if (!selected.ref) return 0;
    const { refList } = this.state;
    const index = refList.findIndex((ref) => ref === selected.ref);
    return index === -1 ? 0 : index;
  };

  onKeyUpEvent = (event) => {
    const { isOpen, refList, selected, focused } = this.state;
    const { key } = event;

    const isEsc = key === "Escape";
    const isClosed = !isOpen;
    const notExpectedKey = !expectedKeys[key];

    const maxIndex = refList.length - 1;
    const firstOptionRef = refList[0];
    const lastOptionRef = refList[maxIndex];

    if (notExpectedKey) return;

    if (isEsc) {
      const focusedIndex = this.getSelectedOptionIndex(selected);
      this.setState({ isOpen: false, focused: { index: focusedIndex } });
      return;
    }

    if (isClosed) {
      const focusedIndex = this.getSelectedOptionIndex(selected);
      this.setState({ isOpen: true, focused: { index: focusedIndex } });
      return;
    }

    let focusedIndex = focused.index;
    let focusedOption;

    switch (key) {
      case "PageDown":
        lastOptionRef.current.focus();
        focusedIndex = maxIndex;
        this.setState({ focused: { index: focusedIndex } });
        break;
      case "PageUp":
        firstOptionRef.current.focus();
        focusedIndex = 0;
        this.setState({ focused: { index: focusedIndex } });
        break;
      case "ArrowDown":
        if (focusedIndex === maxIndex) return;

        focusedIndex += 1;
        focusedOption = refList[focusedIndex];
        focusedOption.current.focus();

        this.setState({ focused: { index: focusedIndex } });
        break;
      case "ArrowUp":
        if (focusedIndex === 0) return;

        focusedIndex -= 1;
        focusedOption = refList[focusedIndex];
        focusedOption.current.focus();

        this.setState({ focused: { index: focusedIndex } });
        break;

      default:
        break;
    }
  };

  onSelectEvent = (option) => {
    const { onSelect } = this.props;
    const { event, ref, index } = option;
    this.setState((prev) => ({
      ...prev,
      isOpen: false,
      selected: { ref, index },
      displayText: event.target.name,
    }));

    onSelect(event);
  };

  onBlurEvent = (event) => {
    const { relatedTarget, currentTarget } = event;

    /** hasNoRelatedTarget helps to avoid blur inside the component */
    const hasNoRelatedTarget = !currentTarget.contains(relatedTarget);

    if (hasNoRelatedTarget) {
      this.setState((prev) => ({
        ...prev,
        isOpen: false,
      }));
    }
  };

  setDefaultOption = (defaultOption) => {
    const { ref, text, index } = defaultOption;
    this.setState({ selected: { ref, index }, displayText: text });
  };

  setOptionsRefs = (ref) =>
    this.setState((prev) => ({ ...prev, refList: [...prev.refList, ref] }));

  setTriggerRef = (ref) => {
    this.triggerRef = ref;
  };

  render() {
    const {
      onClickEvent,
      onSelectEvent,
      onBlurEvent,
      onKeyUpEvent,
      setDefaultOption,
      setTriggerRef,
      setOptionsRefs,
    } = this;
    const { children, color, dropdownType, id } = this.props;
    const { isOpen, displayText, selected } = this.state;

    const isOpenClass = isOpen ? "lab-dropdown__content--is-open" : "";

    const trigger = {
      button: <Button text={displayText} />,
      tag: <DropdownTag text={displayText} color={color} />,
    };

    const renderTrigger = trigger[dropdownType];

    return (
      <div className="lab-dropdown" onBlur={onBlurEvent}>
        <DropdownTrigger
          onClickEvent={onClickEvent}
          onKeyUpEvent={onKeyUpEvent}
          id={id}
          setRef={setTriggerRef}
        >
          {renderTrigger}
        </DropdownTrigger>

        <div
          role="menu"
          className={`lab-dropdown__content lab-dropdown__listbox ${isOpenClass}`}
          aria-labelledby={`menu-button--menu--${id}`}
          tabIndex="-1"
          id={`menu--${id}`}
        >
          {React.Children.map(children, (child, index) => {
            const isDropdownSectionTitle = child.type === DropdownSectionTitle;
            // change it to || child.type == OptionItem...
            const isDropdownItem = child.type === TagItem;

            if (isDropdownSectionTitle)
              return <child.type {...child.props} color={color} />;

            if (isDropdownItem)
              return (
                <DropdownOption
                  onSelectEvent={onSelectEvent}
                  color={color}
                  selectedOption={selected}
                  setDefaultOption={setDefaultOption}
                  onKeyUpEvent={onKeyUpEvent}
                  index={index}
                  id={id}
                  setRef={setOptionsRefs}
                >
                  {child}
                </DropdownOption>
              );

            return null;
          })}
        </div>
      </div>
    );
  }
}
