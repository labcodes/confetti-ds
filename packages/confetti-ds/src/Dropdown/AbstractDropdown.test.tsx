import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import _ from "lodash";

import AbstractDropdown from "./AbstractDropdown";
import DropdownSectionTitle from "./DropdownSectionTitle";
import TagDropdownItem from "./TagDropdownItem";
import TagDropdownTrigger from "./TagDropdownTrigger";

describe("AbstractDropdown", () => {
  it("simulates click on trigger", () => {
    const onSelectMock = jest.fn();
    const wrapper = mount(
      <AbstractDropdown
        dropdownType="tag"
        color="mineral"
        defaultText="Click me"
        onSelect={onSelectMock}
        id="test"
      >
        <div>Just some child</div>
      </AbstractDropdown>
    );

    const trigger = wrapper.find(TagDropdownTrigger).at(0);

    expect(
      wrapper
        .find(".lab-dropdown__content")
        .hasClass("lab-dropdown__content--is-open")
    ).toBeFalsy();

    trigger.simulate("click");

    expect(
      wrapper
        .find(".lab-dropdown__content")
        .hasClass("lab-dropdown__content--is-open")
    ).toBeTruthy();

    trigger.simulate("click");

    expect(
      wrapper
        .find(".lab-dropdown__content")
        .hasClass("lab-dropdown__content--is-open")
    ).toBeFalsy();
  });

  it("simulate blur event", () => {
    const onSelectMock = jest.fn();

    const wrapper = mount(
      <AbstractDropdown
        dropdownType="tag"
        color="mineral"
        defaultText="Click me"
        onSelect={onSelectMock}
        id="test"
      >
        <div>Just some child 1</div>
        <DropdownSectionTitle text="First Section" />
        <TagDropdownItem value="1" text="One" />
        <TagDropdownItem value="2" text="Two" />
        <TagDropdownItem value="3" text="Three" />
        <div>Just some child 2</div>
      </AbstractDropdown>
    );

    const trigger = wrapper.find(TagDropdownTrigger).at(0);

    expect(
      wrapper
        .find(".lab-dropdown__content")
        .hasClass("lab-dropdown__content--is-open")
    ).toBeFalsy();

    trigger.simulate("click");

    expect(
      wrapper
        .find(".lab-dropdown__content")
        .hasClass("lab-dropdown__content--is-open")
    ).toBeTruthy();

    trigger.simulate("blur");

    expect(
      wrapper
        .find(".lab-dropdown__content")
        .hasClass("lab-dropdown__content--is-open")
    ).toBeFalsy();
  });

  it("not expected children", () => {
    const onSelectMock = jest.fn();

    const wrapper = mount(
      <AbstractDropdown
        dropdownType="tag"
        color="mineral"
        defaultText="Click me"
        onSelect={onSelectMock}
        id="test"
      >
        <div>Just some child 1</div>
        <div>Just some child 2</div>
        <div>Just some child 3</div>
        <div>Just some child 4</div>
      </AbstractDropdown>
    );

    const childrenLength = wrapper
      .find(".lab-dropdown__content")
      .children().length;

    expect(childrenLength).toBe(0);
  });

  it("expected children", () => {
    const onSelectMock = jest.fn();

    const wrapper = mount(
      <AbstractDropdown
        dropdownType="tag"
        color="mineral"
        defaultText="Click me"
        onSelect={onSelectMock}
        id="test"
      >
        <div>Just some child 1</div>
        <DropdownSectionTitle text="First Section" />
        <TagDropdownItem value="1" text="One" />
        <TagDropdownItem value="2" text="Two" />
        <TagDropdownItem value="3" text="Three" />
        <div>Just some child 2</div>
      </AbstractDropdown>
    );

    const childrenLength = wrapper
      .find(".lab-dropdown__content")
      .children().length;

    expect(childrenLength).toBe(4);
  });

  it("simulate select random option", () => {
    const expectedValues = [
      { value: "1", text: "One" },
      { value: "2", text: "Two" },
      { value: "3", text: "Three" },
    ];

    const onSelectMock = jest.fn();

    const wrapper = mount(
      <AbstractDropdown
        dropdownType="tag"
        color="mineral"
        defaultText="Click me"
        onSelect={onSelectMock}
        id="test"
      >
        <DropdownSectionTitle text="First Section" />

        {expectedValues.map(({ value, text }) => (
          <TagDropdownItem key={value} value={value} text={text} />
        ))}
      </AbstractDropdown>
    );

    const randomIndex = _.random(0, 2);

    const trigger = wrapper.find(TagDropdownTrigger).at(0);
    expect(trigger.text()).toEqual("Click me");

    trigger.simulate("click");

    const selected = wrapper.find(TagDropdownItem).at(randomIndex).at(0);
    expect(onSelectMock.mock.calls.length).toBe(0);
    selected.simulate("click");

    expect(onSelectMock.mock.calls.length).toBe(1);
    expect(trigger.text()).toEqual(expectedValues[randomIndex].text);
  });

  it("simulate select disabled option", () => {
    const expectedValues = [
      { value: "1", text: "One", disabled: true },
      { value: "2", text: "Two" },
      { value: "3", text: "Three" },
    ];

    const onSelectMock = jest.fn();

    const wrapper = mount(
      <AbstractDropdown
        dropdownType="tag"
        color="mineral"
        defaultText="Click me"
        onSelect={onSelectMock}
        id="test"
      >
        <div>Just some child 1</div>
        <DropdownSectionTitle text="First Section" />

        {expectedValues.map(({ value, text, disabled }) => (
          <TagDropdownItem
            key={value}
            value={value}
            text={text}
            disabled={disabled}
          />
        ))}

        <div>Just some child 2</div>
      </AbstractDropdown>
    );

    const trigger = wrapper.find(TagDropdownTrigger).at(0);

    trigger.simulate("click");

    expect(trigger.text()).toEqual("Click me");

    const disabledSelected = wrapper
      .find(TagDropdownItem)
      .find(".lab-tag--disabled")
      .at(0);

    disabledSelected.simulate("click");
    expect(trigger.text()).toEqual("Click me");
  });

  it("renders with base props", async () => {
    const onSelectMock = jest.fn();

    const renderedComponent = renderer
      .create(
        <AbstractDropdown
          dropdownType="tag"
          defaultText="Click me"
          onSelect={onSelectMock}
          id="test"
        >
          <DropdownSectionTitle text="First Section" />
          <TagDropdownItem value="1" text="One" />
          <TagDropdownItem value="2" text="Two" />
          <TagDropdownItem value="3" text="Three" />
        </AbstractDropdown>
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing color as teal", async () => {
    const onSelectMock = jest.fn();

    const renderedComponent = renderer
      .create(
        <AbstractDropdown
          dropdownType="tag"
          defaultText="Click me"
          color="teal"
          onSelect={onSelectMock}
          id="test"
        >
          <DropdownSectionTitle text="First Section" />
          <TagDropdownItem value="1" text="One" />
          <TagDropdownItem value="2" text="Two" />
          <TagDropdownItem value="3" text="Three" />
        </AbstractDropdown>
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <AbstractDropdown
        dropdownType="tag"
        defaultText="Click me"
        color="teal"
        onSelect={onSelectMock}
        id="test"
      >
        <DropdownSectionTitle text="First Section" />
        <TagDropdownItem value="1" text="One" />
        <TagDropdownItem value="2" text="Two" />
        <TagDropdownItem value="3" text="Three" />
      </AbstractDropdown>
    );
    expect(mountedComponent.find(TagDropdownTrigger).prop("color")).toEqual(
      "teal"
    );
    expect(mountedComponent.find(DropdownSectionTitle).prop("color")).toEqual(
      "teal"
    );
  });

  it("calls props.onClose and props.onOpen when clicks on dropdown trigger", async () => {
    const mockOnClose = jest.fn();
    const mockOnOpen = jest.fn();
    const onSelectMock = jest.fn();
    const id = "test_id";
    const mountDropdown = mount(
      <AbstractDropdown
        dropdownType="tag"
        defaultText="Click me"
        onClose={mockOnClose}
        onOpen={mockOnOpen}
        onSelect={onSelectMock}
        id={id}
      >
        <DropdownSectionTitle text="First Section" />
        <TagDropdownItem value="1" text="One" />
        <TagDropdownItem value="2" text="Two" />
        <TagDropdownItem value="3" text="Three" />
      </AbstractDropdown>
    );

    const trigger = mountDropdown.find(TagDropdownTrigger).at(0);
    expect(mockOnOpen.mock.calls.length).toBe(0);
    expect(mockOnClose.mock.calls.length).toBe(0);

    trigger.simulate("click");
    expect(mockOnOpen.mock.calls.length).toBe(1);
    expect(mockOnClose.mock.calls.length).toBe(0);

    trigger.simulate("click");
    expect(mockOnOpen.mock.calls.length).toBe(1);
    expect(mockOnClose.mock.calls.length).toBe(1);
  });

  it("arrow up and arrow down events", async () => {
    const mockOnClose = jest.fn();
    const mockOnOpen = jest.fn();
    const onSelectMock = jest.fn();
    const id = "test_id";

    // We need this to get the activeElement from document.activeElement due to the update of JSDOM
    // https://github.com/jsdom/jsdom/issues/2586#issuecomment-742593116

    const container = document.createElement("div");
    container.id = "container";
    document.body.appendChild(container);

    const wrapper = mount(
      <AbstractDropdown
        dropdownType="tag"
        defaultText="Click me"
        onClose={mockOnClose}
        onOpen={mockOnOpen}
        onSelect={onSelectMock}
        id={id}
      >
        <DropdownSectionTitle text="First Section" />
        <TagDropdownItem value="1" text="One" />
        <TagDropdownItem value="2" text="Two" />
        <TagDropdownItem value="3" text="Three" />
      </AbstractDropdown>,
      { attachTo: document.querySelector("#container") }
    );

    const trigger = wrapper.find(TagDropdownItem).at(0);

    expect(
      wrapper
        .find(".lab-dropdown__content")
        .hasClass("lab-dropdown__content--is-open")
    ).toBeFalsy();

    trigger.simulate("keydown", { key: "ArrowDown" });

    expect(
      wrapper
        .find(".lab-dropdown__content")
        .hasClass("lab-dropdown__content--is-open")
    ).toBeTruthy();

    const options = wrapper.find(TagDropdownItem);

    /** After open the dropdown */
    expect(document.activeElement).toEqual(options.first().getDOMNode());

    /** Move through the options with the ArrowDown key. */
    trigger.simulate("keydown", { key: "ArrowDown" });
    /** Waiting for the HTML Element to be active before run the tests */
    expect(document.activeElement).toEqual(options.at(1).getDOMNode());

    trigger.simulate("keydown", { key: "ArrowDown" });
    expect(document.activeElement).toEqual(options.at(2).getDOMNode());

    trigger.simulate("keydown", { key: "ArrowDown" });
    expect(document.activeElement).toEqual(options.at(2).getDOMNode());

    /** Move through the options with the ArrowUp key. */
    trigger.simulate("keydown", { key: "ArrowUp" });
    expect(document.activeElement).toEqual(options.at(1).getDOMNode());

    trigger.simulate("keydown", { key: "ArrowUp" });
    expect(document.activeElement).toEqual(options.at(0).getDOMNode());

    trigger.simulate("keydown", { key: "ArrowUp" });
    expect(document.activeElement).toEqual(options.at(0).getDOMNode());

    /** Move through the options with the PageDown  key. */
    trigger.simulate("keydown", { key: "PageDown" });
    expect(document.activeElement).toEqual(options.last().getDOMNode());

    /** Move through the options with the PageUp  key. */
    trigger.simulate("keydown", { key: "PageUp" });
    expect(document.activeElement).toEqual(options.first().getDOMNode());
  });
});
