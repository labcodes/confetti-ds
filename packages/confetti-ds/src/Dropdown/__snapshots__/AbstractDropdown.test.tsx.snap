// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbstractDropdown renders as expected when passing color as teal 1`] = `
<div
  className="lab-dropdown"
  onBlur={[Function]}
>
  <button
    aria-selected={false}
    className="lab-tag lab-tag--dropdown lab-tag--outline lab-tag--teal-pale"
    disabled={false}
    id=""
    onClick={[Function]}
    onKeyDown={[Function]}
    role="listbox"
    type="button"
    value=""
  >
    Click me
    <span
      className="lab-tag__dropdown-icon-wrapper"
    >
      <span
        className="lab-icon lab-icon--dropdown-closed lab-tag--dropdown-icon"
      >
        <svg
          aria-hidden={true}
          className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
          fill="currentColor"
          height={20}
          viewBox="0 0 20 20"
          width={20}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.02 8.63A1 1 0 0 1 5.8 7h8.4a1 1 0 0 1 .78 1.63l-3.81 4.72a1.5 1.5 0 0 1-2.34 0L5.02 8.63ZM14.2 8H5.8l3.81 4.72c.2.25.58.25.78 0L14.2 8Z"
            fill="currentColor"
          />
        </svg>
      </span>
    </span>
  </button>
  <div
    aria-labelledby="menu-button--menu--test"
    className="lab-dropdown__content lab-dropdown__listbox "
    id="menu--test"
    role="menu"
    tabIndex={-1}
  >
    <span
      className="lab-dropdown__section-title lab-dropdown__section-title--color-teal"
    >
      First Section
    </span>
    <div
      className="lab-dropdown__option "
      id="lab-dropdown__option--1"
    >
      <button
        aria-selected={false}
        className="lab-tag lab-tag--togglable lab-tag--pale"
        disabled={false}
        id="1"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="option"
        tabIndex="-1"
        type="button"
        value="1"
      >
        One
      </button>
    </div>
    <div
      className="lab-dropdown__option "
      id="lab-dropdown__option--2"
    >
      <button
        aria-selected={false}
        className="lab-tag lab-tag--togglable lab-tag--pale"
        disabled={false}
        id="2"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="option"
        tabIndex="-1"
        type="button"
        value="2"
      >
        Two
      </button>
    </div>
    <div
      className="lab-dropdown__option "
      id="lab-dropdown__option--3"
    >
      <button
        aria-selected={false}
        className="lab-tag lab-tag--togglable lab-tag--pale"
        disabled={false}
        id="3"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="option"
        tabIndex="-1"
        type="button"
        value="3"
      >
        Three
      </button>
    </div>
  </div>
</div>
`;

exports[`AbstractDropdown renders with base props 1`] = `
<div
  className="lab-dropdown"
  onBlur={[Function]}
>
  <button
    aria-selected={false}
    className="lab-tag lab-tag--dropdown lab-tag--outline lab-tag--teal-pale"
    disabled={false}
    id=""
    onClick={[Function]}
    onKeyDown={[Function]}
    role="listbox"
    type="button"
    value=""
  >
    Click me
    <span
      className="lab-tag__dropdown-icon-wrapper"
    >
      <span
        className="lab-icon lab-icon--dropdown-closed lab-tag--dropdown-icon"
      >
        <svg
          aria-hidden={true}
          className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
          fill="currentColor"
          height={20}
          viewBox="0 0 20 20"
          width={20}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.02 8.63A1 1 0 0 1 5.8 7h8.4a1 1 0 0 1 .78 1.63l-3.81 4.72a1.5 1.5 0 0 1-2.34 0L5.02 8.63ZM14.2 8H5.8l3.81 4.72c.2.25.58.25.78 0L14.2 8Z"
            fill="currentColor"
          />
        </svg>
      </span>
    </span>
  </button>
  <div
    aria-labelledby="menu-button--menu--test"
    className="lab-dropdown__content lab-dropdown__listbox "
    id="menu--test"
    role="menu"
    tabIndex={-1}
  >
    <span
      className="lab-dropdown__section-title lab-dropdown__section-title--color-teal"
    >
      First Section
    </span>
    <div
      className="lab-dropdown__option "
      id="lab-dropdown__option--1"
    >
      <button
        aria-selected={false}
        className="lab-tag lab-tag--togglable lab-tag--pale"
        disabled={false}
        id="1"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="option"
        tabIndex="-1"
        type="button"
        value="1"
      >
        One
      </button>
    </div>
    <div
      className="lab-dropdown__option "
      id="lab-dropdown__option--2"
    >
      <button
        aria-selected={false}
        className="lab-tag lab-tag--togglable lab-tag--pale"
        disabled={false}
        id="2"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="option"
        tabIndex="-1"
        type="button"
        value="2"
      >
        Two
      </button>
    </div>
    <div
      className="lab-dropdown__option "
      id="lab-dropdown__option--3"
    >
      <button
        aria-selected={false}
        className="lab-tag lab-tag--togglable lab-tag--pale"
        disabled={false}
        id="3"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="option"
        tabIndex="-1"
        type="button"
        value="3"
      >
        Three
      </button>
    </div>
  </div>
</div>
`;
